### PR TITLE
Feature - Send-Via / TransferSender to send messages across entities in a transaction

### LINKED ISSUE #423: Send-Via feature is missing
With [Transactions support](https://github.com/Azure/azure-service-bus-dotnet/issues/6) in place (PR #421), `send-via` is missing to allow trasport transaction with source and destination be on the different queues (see https://github.com/Azure/azure-service-bus-dotnet/issues/6#issuecomment-366082384).

### PR DESCRIPTION
**Pull request by nemakam**
## Description
This feature lets a user send messages across entities in a single transaction.

A single transaction cannot theoretically span across entities. So as to support cross-entity transaction, you can now send a message to a `destination-queue` via another queue. The transaction will be performed on the `via-queue`, and once successful, the message will be forwarded/transferred to its intended destination.
Resolves #423 

Sample:
```csharp
var viaQueueSender = new MessageSender(connection, viaQueueName);
var viaQueueReceiver = new MessageReceiver(connection, viaQueueName);
var destinationViaSender = new MessageSender(connection, destinationQueueName, viaQueueName);

var message1 = new Message(body) { MessageId = "1", PartitionKey = "pk1" };
var message2 = new Message(body) { MessageId = "2", PartitionKey = "pk2", ViaPartitionKey = "pk1" };

await viaQueueSender.SendAsync(message1);
var receivedMessage = await viaQueueReceiver.ReceiveAsync();

using (new TransactionScope(TransactionScopeAsyncFlowOption.Enabled))
{
	await viaQueueReceiver.CompleteAsync(receivedMessage.SystemProperties.LockToken);
	await destinationViaSender.SendAsync(message2);
}
```

In the above example, operations on both `message1` and `message2` (intended for different entities) are being performed within the same entity (i.e., the `via-queue`)

### CODE REVIEW

File: src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs [hunk 1]
```diff
@@ -252,6 +284,11 @@ public Task SendAsync(Message message)
                     "Cannot schedule messages in the past");
             }
 
+            if (this.isViaSender && Transaction.Current != null)
+            {
+                throw new ServiceBusException(false, $"{nameof(ScheduleMessageAsync)} method is not supported in a Via-Sender.");
```
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs]: Should exception message be more specific as per _why_ it's not supported? There are two constraints, send-via and transaction.

**Comment by SeanFeldman** [comment]
@nemakam while waiting for @binzywu's review, should the test on .NET Standard be addressed?

**Comment by nemakam** [comment]
@SeanFeldman 
This requires service side changes that are yet to be deployed. The send-via test will fail meanwhile. And once a test fails, it drags down other tests as well as the queue is reused. 
That's why I've put the 'onhold' tag :)

**Comment by SeanFeldman** [comment]
We need a `patience-sean` tag 😆

**Comment by nemakam** [comment]
**nemakam** added **patience-sean** label 5 minutes ago

**Comment by SeanFeldman** [comment]
Epic! 😂

**Comment by SeanFeldman** [comment]
@nemakam any updates? Thanks.

**Comment by nemakam** [comment]
@SeanFeldman 
ETA - for preview build - a week or so. We are still waiting for the release to finish.
Then once internal long running tests pass, we can open it up. Needs couple of days for this as well.

**Pull request merged**